src/Microsoft.Azure.WebJobs.Extensions.EventHubs/Listeners/EventHubListener.cs

```diff
@@ -71,21 +75,34 @@ void IListener.Cancel()
 
         void IDisposable.Dispose()
         {
+            StopAsync(CancellationToken.None).Wait();
         }
 
         public async Task StartAsync(CancellationToken cancellationToken)
         {
             await _eventProcessorHost.RegisterEventProcessorFactoryAsync(this, _options.EventProcessorOptions);
             _started = true;
+
+            _logger.LogDebug($"EventHub listener started ({_details})");
         }
 
         public async Task StopAsync(CancellationToken cancellationToken)
         {
-            if (_started)
+            await _stopSemaphoreSlim.WaitAsync();
+            try
+            {
+                if (_started)
+                {
+                    await _eventProcessorHost.UnregisterEventProcessorAsync();
+                }
+                _started = false;
+            }
+            finally
             {
-                await _eventProcessorHost.UnregisterEventProcessorAsync();
+                _stopSemaphoreSlim.Release();
             }
-            _started = false;
+
+            _logger.LogDebug($"EventHub listener stopped ({_details})");

```

Comment:
Move this within the if (_started) block so we don't log on a no-op call to StopAsync() (e.g. when StopAsync is called for the second time)

Reply:
Alternately, its fine if we want to show how many times StopAsync() was called.  But in this case, we may want to log "stopped" in the if (_started) block and a different log statement in the else condition to indicate no action was needed as the listener is already stopped.